components/eamxx/src/physics/shoc/eamxx_shoc_process_interface.cpp

```diff
@@ -461,5 +465,53 @@ void SHOCMacrophysics::finalize_impl()
   // Do nothing
 }
 // =========================================================================================
+void SHOCMacrophysics::check_flux_state_consistency(const double dt)
+{
+  using PC = scream::physics::Constants<Real>;
+  const Real gravit = PC::gravit;
+  const Real qmin   = 1e-12; // minimum permitted constituent concentration (kg/kg)
+
+  const auto& pseudo_density = get_field_in ("pseudo_density").get_view<const Spack**>();
+  const auto& surf_evap      = get_field_out("surf_evap").get_view<Real*>();
+  const auto& qv             = get_field_out("qv").get_view<Spack**>();
 
+  const auto nlevs           = m_num_levs;
+  const auto nlev_packs      = ekat::npack<Spack>(nlevs);
+  const auto last_pack_idx   = (nlevs-1)/Spack::n;
+  const auto last_pack_entry = (nlevs-1)%Spack::n;
+  const auto policy          = ekat::ExeSpaceUtils<KT::ExeSpace>::get_default_team_policy(m_num_cols, nlev_packs);
+  Kokkos::parallel_for("check_flux_state_consistency",
+                       policy,
+                       KOKKOS_LAMBDA (const KT::MemberType& team) {
+    const auto i = team.league_rank();
+
+    const auto& pseudo_density_i = ekat::subview(pseudo_density, i);
+    const auto& qv_i             = ekat::subview(qv, i);
+
+    // reciprocal of pseudo_density at the bottom layer
+    const auto rpdel = 1.0/pseudo_density_i(last_pack_idx)[last_pack_entry];
+
+    // Check if the negative surface latent heat flux can exhaust
+    // the moisture in the lowest model level. If so, apply fixer.
+    const auto condition = surf_evap(i) - (qmin - qv_i(last_pack_idx)[last_pack_entry])/(dt*gravit*rpdel);
+    if (condition < 0) {
+      const auto cc = abs(surf_evap(i)*dt*gravit);
+
+      auto tracer_mass = [&](const int k) {
+        return qv_i(k)*pseudo_density_i(k);
+      };
+      Real mm = ekat::ExeSpaceUtils<KT::ExeSpace>::view_reduction(team, 0, nlevs, tracer_mass);
+
+      EKAT_KERNEL_ASSERT_MSG(mm >= cc, "Error! Total mass of column vapor should be greater than mass of surf_evap.\n");
+
+      Kokkos::parallel_for(Kokkos::TeamVectorRange(team, nlev_packs), [&](const int& k) {
+        const auto adjust = cc*qv_i(k)*pseudo_density_i(k)/mm;
+        qv_i(k) = (qv_i(k)*pseudo_density_i(k) - adjust)/pseudo_density_i(k);
+      });
+
+      surf_evap(i) = 0;
+    }
+  });
```

Comment:
i am confused about this code:

1. i think it is being called in each macmic iteration, which means only surf_evap/macmic_loop_size is applied in shoc iteration. does the code operate with quantity surf_evap/macmic_loop_size or does it operate with surf_evap?
2. the way i read the code, compared to the original version (before 'condition' line was introduced) is that nothing in the distribution of extra negative mass changed. i think Peter's objection was that if there is potentially negative mass in the bottom layer, then the excess would be redistributed above it. the way i read the code, right now not excess (value connected to 'condition'), but the whole surf_evap mass (value connected to 'cc') is redistributed, and not to the top of the very bottom layer, but to all layers (parallel_for loop).
3. related to that, i think the algorithm that Peter B. wanted to see would leave surf_evap not with 0, but with the biggest by the abs value mass that is allowed in the bottom layer.